### DISCUSSION TITLE
DAI 2024-2025 - Java IOs benchmarking - Basile Buxtorf

### DISCUSSION THREAD

**Discussion by BasileBux** [heig-vd-dai-course/heig-vd-dai-course#368] [Show and tell]
https://github.com/BasileBux/heig-vd-dai-course-java-ios

### Binary
| Size    | Writer | Reader | Buffer writer | Buffer reader |
| ------- | ------ | ------ | ------------- | ------------- |
| 1B       | 214    | 201    | 213           | 208           |
| 1KiB    | 251    | 218    | 220           | 253           |
| 1MiB | 6367   | 1580   | 319           | 261           |
| 5MiB | 31313  | 6985   | 357           | 469           |

### Text

| Size    | Writer | Reader | Buffer writer | Buffer reader |
| ------- | ------ | ------ | ------------- | ------------- |
| 1B       | 242    | 213    | 231           | 211           |
| 1KiB    | 226    | 223    | 243           | 204           |
| 1MiB | 656    | 354    | 351           | 276           |
| 5MiB | 1073   | 678    | 494           | 419           |

In lower sizes, both streams are really comparable in time. However, as sizes get bigger, the buffered streams scale extremely well which isn't the case for non-buffered streams. 

Buffered streams are more efficient because they reduce the number of I/O operations. Instead of processing data byte by byte, they gather data into a buffer and write or read it in larger chunks. This minimizes the overhead caused by frequent disk I/O access, which is especially costly for larger files.

There isn't any difference between text and binary in the files directly. All files are binary, the only difference is the interpretation and how the computer reads them. And for text files, there is character encoding which is specific to text files. 

Character encoding is a system that maps characters (letters, symbols, etc.) to specific byte sequences. It ensures that text data is stored and transmitted properly and that the same data can be interpreted correctly across different platforms and systems.

Benchmark methodology is really important because, as we all use slightly different configurations, the benchmark should be applicable to most system and the final observation should generalize to all (hardware / software) configurations.

**Comment by ludelafo** [comment]
Thank you!

You can always find the solution in the solution repository if you ever need to compare your work with the official one here: https://github.com/heig-vd-dai-course/heig-vd-dai-course-solutions.

Do not hesitate to ask questions if something is not clear, either in this discussion or in the Q&A category of the GitHub Discussions here: https://github.com/orgs/heig-vd-dai-course/discussions/categories/q-a.
